### ISSUE TITLE
Attention Layer

### ISSUE THREAD

**Issue by shuagr** [lucidrains/DALLE-pytorch#30]
How are you doing the sparse attention mentioned in the blog?
"The attention mask at each of its 64 self-attention layers allows each image token to attend to all text tokens. DALL·E uses the standard causal mask for the text tokens, and sparse attention for the image tokens with either a **row, column, or convolutional attention pattern**, depending on the layer. "

You are alternating between dense and sparse attention instead of alternating between all sparse. 

Can you explain that part of the code?

Also, how are you padding the text if input text < 256 tokens.? Are you changing sparse attention for each input? or how are you masking text tokens with different padding for left to right generation with the input of different text in a multi-batch scenario

**Comment by lucidrains** [comment]
@shuagr Hello! As of today, this repository only offers either full causal attention, or the specific sparse attention that comes with Deepspeed. The row and column attention patterns likely come from this paper https://arxiv.org/abs/1912.12180 , and the sparse convolution attention pattern is completely new to me. I am guessing they have someone who could craft the special CUDA kernels for the latter.

However, given https://openai.com/blog/sparse-transformer/ , I am sure the Deepspeed sparse attention and/or full attention will be enough to reproduce results.

Re: text, padding and attention. I am making the assumption of 0 as padding, and one could supply an extra mask `batch x seq`, which will appropriately mask out the attention to the padding for both full and the sparse attention variant.

**Comment by lucidrains** [comment]
@shuagr I will likely bring in some other sparse attention techniques eventually, specifically https://github.com/lucidrains/sinkhorn-transformer and https://github.com/lucidrains/performer-pytorch

**Comment by lucidrains** [comment]
Btw I think I figured out the autoregressive axial and convolutional sparse attention they are doing. I'll be adding the two custom sparse attention classes to the repo tomorrow

**Comment by lucidrains** [comment]
@shuagr I think I figured out how to do the sparse causal convolution that OpenAI mentioned in their blogpost, all without CUDA code

https://github.com/lucidrains/DALLE-pytorch/blob/0.0.50/dalle_pytorch/attention.py#L74-L157

Could use a code review!

**Comment by lucidrains** [comment]
@shuagr ok, i completed the axial attention too, as I imagined it from reading the blog post

https://github.com/lucidrains/DALLE-pytorch/releases/tag/0.0.51

That should complete all the possible attention

Let me know if that looks good and I'll wire it into the transformers so that one can specify any type of attention on any given depth

**Comment by lucidrains** [comment]
https://github.com/lucidrains/DALLE-pytorch/blob/main/dalle_pytorch/attention.py#L162-L243 Sparse Axial Attention

**Comment by lucidrains** [comment]
@shuagr  https://github.com/lucidrains/DALLE-pytorch#sparse-attention feature complete!
